cmd/gateway/s3x/crdt_dagsyncer.go

```diff
@@ -0,0 +1,122 @@
+package s3x
+
+import (
+	"context"
+
+	pb "github.com/RTradeLtd/TxPB/v3/go"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+)
+
+//crdtDAGSyncer implements crdt.DAGSyncer using a pb.NodeAPIClient and a datastore
+type crdtDAGSyncer struct {
+	client pb.NodeAPIClient
+	ds     datastore.Batching
+}
+
+// Get retrieves nodes by CID. Depending on the NodeGetter
+// implementation, this may involve fetching the Node from a remote
+// machine; consider setting a deadline in the context.
+func (d *crdtDAGSyncer) Get(ctx context.Context, c cid.Cid) (ipld.Node, error) {
+	resp, err := d.client.Dag(ctx, &pb.DagRequest{
+		RequestType: pb.DAGREQTYPE_DAG_GET,
+		Hash:        c.String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	block := blocks.NewBlock(resp.RawData)
+	if block.Cid() != c {
+		return nil, errors.New("unexpected data received from node server")
+	}
+	n, err := ipld.Decode(block)
+	return n, d.setBlock(c, err)
+}
+
+// GetMany returns a channel of NodeOptions given a set of CIDs.
+func (d *crdtDAGSyncer) GetMany(ctx context.Context, cs []cid.Cid) <-chan *ipld.NodeOption {
+	out := make(chan *ipld.NodeOption, len(cs))
+	go func() {
+		for _, c := range cs {
+			n, err := d.Get(ctx, c)
+			out <- &ipld.NodeOption{
+				Node: n,
+				Err:  err,
+			}
+		}
+		close(out)
+	}()
+	return out
+}
+
+// Add adds a node to this DAG.
+func (d *crdtDAGSyncer) Add(ctx context.Context, n ipld.Node) error {
+	return d.AddMany(ctx, []ipld.Node{n})
+}
+
+// AddMany adds many nodes to this DAG.
+//
+// Consider using the Batch NodeAdder (`NewBatch`) if you make
+// extensive use of this function.
+func (d *crdtDAGSyncer) AddMany(ctx context.Context, ns []ipld.Node) error {
+	cs := make([]string, 0, len(ns))
+	for _, n := range ns {
+		switch typed := n.(type) {
+		default:
+			return errors.Errorf("Can not add type: %T using dag client", n)
+		case *merkledag.ProtoNode:
+			_, err := ipfsSaveProtoNode(ctx, d.client, typed)
+			if err != nil {
+				return errors.WithMessage(err, "error decoding returned cid")
+			}
+		}
+		if err := d.setBlock(n.Cid()); err != nil {
+			return err
+		}
+		cs = append(cs, n.Cid().String())
+	}
+	_, err := d.client.Persist(ctx, &pb.PersistRequest{Cids: cs}) //Question: should persist be requested before add?
+	if err != nil {
+		return err
+	}
+	return nil
```

Comment:
can probably just condense this into:

````Go
_, err = d.client.Persist(ctx, &pb.PersistRequest{Cids: cds})
return err
```

Also what do you mean by this:

```
//Question: should persist be requested before add?
```

Reply:
Actually, I'm not too sure if `Persist` is necessery here at all.

Reply:
I dont think it is, `Persist` is the equivalent to `ipfs pin add`, and is intended to get data from the network by CID and store it locally. In this cale `ipfsSaveProtoNode` will store the data, and as such persist isn't needed